middleware/assert_test.go

```diff
@@ -14,106 +14,149 @@ import (
-	Convey("Given a healthy dp-cantabular-filter-flex-api", t, func() {
-		expectedResponse := &http.Response{
-			Body:       io.NopCloser(bytes.NewReader([]byte(testBody))),
-			StatusCode: http.StatusOK,
-			Header: map[string][]string{
-				"X-Test": []string{"Value"},
+func setupFilterFlexAPIMock() *apimock.FilterFlexAPIMock {
+	expectedResponse := &http.Response{
+		Body:       io.NopCloser(bytes.NewReader([]byte(testBody))),
+		StatusCode: http.StatusOK,
+		Header: map[string][]string{
+			"X-Test": []string{"Value"},
+		},
+	}
+
+	return &apimock.FilterFlexAPIMock{
+		ForwardRequestFunc: func(r *http.Request) (*http.Response, error) {
+			return expectedResponse, nil
+		},
+	}
+}
+
+func setupDataStoreMock(filterIsFlexible bool, testBluePrintID string) *apimock.DataStoreMock {
+	datastoreMock := mock.NewDataStore().Mock
+	datastoreMock.GetFilterOutputFunc = func(ctx context.Context, filterOutputID string) (*models.Filter, error) {
+		return &models.Filter{
+			Dataset: &models.Dataset{
+				Version: 1,
+			},
+			ID: filterOutputID,
+			Links: models.LinkMap{
+				FilterBlueprint: &models.LinkObject{ID: testBluePrintID, HRef: ""},
 			},
+		}, nil
+	}
+	datastoreMock.GetFilterFunc = func(ctx context.Context, filterID, etag string) (*models.Filter, error) {
+		filter := &models.Filter{
+			Dataset: &models.Dataset{
+				Version: 1,
+			},
+			ID: filterID,
+		}
+		if filterIsFlexible {
+			filter.Type = "flexible"
 		}
+		return filter, nil
+	}
+	return datastoreMock
+}
 
-		filterFlexAPIMock := &apimock.FilterFlexAPIMock{
-			ForwardRequestFunc: func(r *http.Request) (*http.Response, error) {
-				return expectedResponse, nil
-			},
+func setupDatasetAPIMock(datasetIsFlexible bool) *apimock.DatasetAPIMock {
+	datasetAPIMock := &apimock.DatasetAPIMock{}
+
+	datasetAPIMock.GetFunc = func(ctx context.Context, ut, st, cid, dsid string) (dataset.DatasetDetails, error) {
+		dataset := dataset.DatasetDetails{
+			Type: "not_flexible",
 		}
+		if datasetIsFlexible {
+			dataset.Type = "cantabular_flexible_table"
+		}
+		return dataset, nil
+	}
+	return datasetAPIMock
+}
 
-		Convey("When a filter with given ID in the datastore of type 'flexible' is found", func(){
-			datastoreMock := mock.NewDataStore().Mock
-			datastoreMock.GetFilterFunc = func(ctx context.Context, filterID, etag string) (*models.Filter, error) {
-				return &models.Filter{
-					Type: "flexible",
-					Dataset: &models.Dataset{
-						Version: 1,
-					},
-					ID: filterID,
-				}, nil
-			}
-
-			Convey("When an incoming request passes through the assert.FilterType middleware", func(){
-				assert := middleware.NewAssert(
-					responder.New(),
-					&mock.DatasetAPI{},
-					filterFlexAPIMock,
-					datastoreMock,
-					testToken,
-					true,
-				)
+func assertTestSetup(filterIsFlexible, datasetIsFlexible, assertIsEnabled bool, testBluePrintID string) (*apimock.DataStoreMock, *apimock.FilterFlexAPIMock, *apimock.DatasetAPIMock, *middleware.Assert) {
```

Comment:
👍

Might even be able to take this further and have it run the req, although doesn't really save that many lines:

```go
func runassertTest(req *http.Request, filterIsFlexible, datasetIsFlexible, assertIsEnabled bool, testBluePrintID string) (*apimock.DataStoreMock, *apimock.FilterFlexAPIMock, *apimock.DatasetAPIMock, *httptest.ResponseRecorder) {
	filterFlexAPIMock := setupFilterFlexAPIMock()
	datastoreMock := setupDataStoreMock(filterIsFlexible, testBluePrintID)
	datasetAPIMock := setupDatasetAPIMock(datasetIsFlexible)

	assert := middleware.NewAssert(
		responder.New(),
		datasetAPIMock,
		filterFlexAPIMock,
		datastoreMock,
		testToken,
		assertIsEnabled,
	)

	w := httptest.NewRecorder()
	next := http.HandlerFunc(testHandler)
	f := assert.FilterType(next)
	f.ServeHTTP(w, req)

	return datastoreMock, filterFlexAPIMock, datasetAPIMock, w
}
```

Reply:
Actually ignore me, because the method will be different.